src/app-e-heat-input-from-oil-workspace/app-e-heat-input-from-oil.service.ts

```diff
@@ -81,29 +84,55 @@ export class AppEHeatInputFromOilWorkspaceService {
       );
     }
 
-    let entity = this.repository.create({
-      id: historicalRecordId ? historicalRecordId : uuid(),
-      monitoringSystemId: system.id,
-      oilMass: payload.oilMass,
-      oilGCV: payload.oilGCV,
-      oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
-      oilHeatInput: payload.oilHeatInput,
-      oilVolume: payload.oilVolume,
-      oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
-      oilDensity: payload.oilDensity,
-      oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
-      appECorrTestRunId,
-      userId,
-      addDate: timestamp,
-      updateDate: timestamp,
-    });
+    let entity;
+
+    if (trx) {
+      entity = trx.getRepository(AppEHeatInputFromOil).create({
+        id: historicalRecordId ? historicalRecordId : uuid(),
+        monitoringSystemId: system.id,
+        oilMass: payload.oilMass,
+        oilGCV: payload.oilGCV,
+        oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
+        oilHeatInput: payload.oilHeatInput,
+        oilVolume: payload.oilVolume,
+        oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
+        oilDensity: payload.oilDensity,
+        oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
+        appECorrTestRunId,
+        userId,
+        addDate: timestamp,
+        updateDate: timestamp,
+      });
+
+      await trx.getRepository(AppEHeatInputFromOil).save(entity);
+      entity = await trx.getRepository(AppEHeatInputFromOil).findOneBy({id: entity.id});
+    } else {
+      entity = this.repository.create({
+        id: historicalRecordId ? historicalRecordId : uuid(),
+        monitoringSystemId: system.id,
+        oilMass: payload.oilMass,
+        oilGCV: payload.oilGCV,
+        oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
+        oilHeatInput: payload.oilHeatInput,
+        oilVolume: payload.oilVolume,
+        oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
+        oilDensity: payload.oilDensity,
+        oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
+        appECorrTestRunId,
+        userId,
+        addDate: timestamp,
+        updateDate: timestamp,
+      });
+
+      await this.repository.save(entity);
+      entity = await this.repository.getAppEHeatInputFromOilById(entity.id);
+    }
```

Comment:
This should use the `withTransaction` method (described in another comment) so that the `getAppEHeatInputFromOilById` method can be used even when in a transaction:
```suggestion
    const repository = withTransaction(this.repository, trx);
    entity = repository.create({
      id: historicalRecordId ? historicalRecordId : uuid(),
      monitoringSystemId: system.id,
      oilMass: payload.oilMass,
      oilGCV: payload.oilGCV,
      oilGCVUnitsOfMeasureCode: payload.oilGCVUnitsOfMeasureCode,
      oilHeatInput: payload.oilHeatInput,
      oilVolume: payload.oilVolume,
      oilVolumeUnitsOfMeasureCode: payload.oilVolumeUnitsOfMeasureCode,
      oilDensity: payload.oilDensity,
      oilDensityUnitsOfMeasureCode: payload.oilDensityUnitsOfMeasureCode,
      appECorrTestRunId,
      userId,
      addDate: timestamp,
      updateDate: timestamp,
    });

    await repository.save(entity);
    entity = await repository.getAppEHeatInputFromOilById(entity.id);
```